<a href="https://github.com/fgonzaleumbc/DATA601/blob/main/Homework/Homework_4/DATA601_Spring23_HW_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# STUDENT NAME AND INFORMATION

Note: For the cell above, feel free to use the markdown code from the class Jupyter Notebook and modify it with your information. 

#### Due-date: Monday 3/13/2023 midnight

Please Save a copy of this notebook to modify and submit your homework. Rename the file accordingly. If using Colab, "Open in Colab" link on the left top of this notebook to save a copy of this notebook to your "Colab Notebooks" under your Google Drive, and rename it. 

Please provide your solutions cells after each question. You can create new cells as needed. <br>

Please return your notebook by the due date via by emailing the grader and CC instructor as a *.ipynb file) via email or Blackboard. The name of your notebook should be in the following format: YourLastname_HW01.ipynb (e.g. Gonzalez_HW01.ipyn).

Make sure that your code is in a working condition when you restart the kernel. Notebooks are great tools for exploration but we should be very careful as we are writing codes.

# Dataset Source: 

This HW will use the Titanic dataset which is a common dataset used in data science. Please use the file "titanic.csv" in the input_data folder.

Citation:

The original Titanic dataset, describing the survival status of individual passengers on the Titanic. The titanic data does not contain information from the crew, but it does contain actual ages of half of the passengers. The principal source for data about Titanic passengers is the Encyclopedia Titanica. The datasets used here were begun by a variety of researchers. One of the original sources is Eaton & Haas (1994) Titanic: Triumph and Tragedy, Patrick Stephens Ltd, which includes a passenger list created by many researchers and edited by Michael A. Findlay.

Thomas Cason of UVa has greatly updated and improved the titanic data frame using the Encyclopedia Titanica and created the dataset here. Some duplicate passengers have been dropped, many errors corrected, many missing ages filled in, and new variables created.

Attribute information
The variables on our extracted dataset are pclass, survived, name, age, embarked, home.dest, room, ticket, boat, and sex. pclass refers to passenger class (1st, 2nd, 3rd), and is a proxy for socio-economic class. Age is in years, and some infants had fractional values. The titanic2 data frame has no missing data and includes records for the crew, but age is dichotomized at adult vs. child. These data were obtained from Robert Dawson, Saint Mary's University, E-mail. The variables are pclass, age, sex, survived. These data frames are useful for demonstrating many of the functions in Hmisc as well as demonstrating binary logistic regression analysis using the Design library. For more details and references see Simonoff, Jeffrey S (1997): The "unusual episode" and a second statistics course. J Statistics Education, Vol. 5 No. 1.

For more information about how this dataset was constructed: http://biostat.mc.vanderbilt.edu/wiki/pub/Main/DataSets/titanic3info.txt

Other Titanic Dataset Information Sources:
- https://www.openml.org/search?type=data&sort=runs&id=40945&status=active
- https://www.kaggle.com/competitions/titanic/data?select=gender_submission.csv

# Homework Questions:

#### Question 1 (3 points): Load file and name it "df_titanic_1".

#### Question 2 (3 points): Provide a sample of the dataframe (i.e., the output should be at least 5 rows of data with the columns).

#### Question 3 (4 points): Calculate number of rows and of columns.

#### Question 4 (3 points): Show the datatypes of each column. 

#### Question 5 (3 points): Provide some discussion and thoughts on the column cabin. Does it provide any useful information? Should it be removed? Should it be kept? 
*Note: Question 6 is irrelevant to this question. 

#### Question 6 (3 points): Remove Colum "cabin" from df_titanic_1.

#### Question 7 (3 points): Create a dataframe called df_passenger_info whcih includes columns "passenger", "name", "sex", and "age".

#### Question 8 (3 points): Convert the df_passenger_info to a dictionary.

#### Question 9 (5 points): Create a dataframe called "df_titanic_2" that is filtering for passengers that paid a fare higher than 10 AND have an age less than 50.

#### Question 10 (5 points): Sort df_titanic_2 by 'age' column in ascending order.

#### Question 11 (5 points): Export the "df_titanic_2" as a csv and put in output_data folder

#### Question 12 (5 points): Export the "df_passenger_info" as a JSON put in output_data folder

# Homework Answers:

In [1]:
import pandas as pd
import numpy as np

In [3]:
#Loading the titanic data

df_titanic_1= pd.read_csv('./input_data/titanic.csv')

In [4]:
#Displays the dataframe with 5 rows

df_titanic_1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [5]:
#Displays rows and columns in the dataframe

print("number of rows", df_titanic_1.shape[0], '\n', "number of columns", df_titanic_1.shape[1])

number of rows 891 
 number of columns 12


In [6]:
#displays the datatypes in the dataframe

print('the dataypes in this dataframe are:')
df_titanic_1.dtypes

the dataypes in this dataframe are:


PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [7]:
#displays the sum of missing values with respect to each column in the dataframe

df_titanic_1.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

From the above, it is clear that out of 891 records 687 are missing for cabin column, which is almost 77% of missingness. Hence, it better to remove that column from the dataframe

In [8]:
# Drops the column cabin

df_titanic_1.drop(columns= 'Cabin',axis=1, inplace= True)
df_titanic_1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [9]:
# Creating a new dataframe with required columns

df_passenger_info= df_titanic_1[['PassengerId','Name','Sex','Age']]

df_passenger_info.head()

PassengerId                                               Name     Sex  \
0            1                            Braund, Mr. Owen Harris    male   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2            3                             Heikkinen, Miss. Laina  female   
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4            5                           Allen, Mr. William Henry    male   

    Age  
0  22.0  
1  38.0  
2  26.0  
3  35.0  
4  35.0

In [15]:
# Converting the dataframe to dictionary

df_passenger_info_dict = df_passenger_info.to_dict(orient= 'records')
df_passenger_info_dict[:3] # Showing a sample only (first three records).

[{'PassengerId': 1,
  'Name': 'Braund, Mr. Owen Harris',
  'Sex': 'male',
  'Age': 22.0},
 {'PassengerId': 2,
  'Name': 'Cumings, Mrs. John Bradley (Florence Briggs Thayer)',
  'Sex': 'female',
  'Age': 38.0},
 {'PassengerId': 3,
  'Name': 'Heikkinen, Miss. Laina',
  'Sex': 'female',
  'Age': 26.0}]

In [22]:
# Filtering dataframe based on conditions

df_titanic_2= df_titanic_1[(df_titanic_1['Fare']> 10) & (df_titanic_1['Age'] < 50)].copy()
# Note: if .copy() is not used,and will cause warning message when modifying df_titanic_2 due to referencing df_titanic_1.
df_titanic_2

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
883          884         0       2   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
883                      Banfield, Mr. Frederick James    male  28.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Embarked  
1        0          PC 17599  71.2833        C  
3        0            113803  53.1000        S  
7        1            349909  21.0750        S  
8        2            347742  11.1333        S  
9        0            237736  30.0708        C  
..     ...               ...      ...      ...  
883      0  C.A./SOTON 34068  10.5000        S  
885      5            382652  29.1250        Q  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
889      0            111369  30.0000        C  

[415 rows x 11 columns]

In [23]:
# Sorting the dataframe based on age column in ascending order

df_titanic_2.sort_values(by='Age', ascending= True, inplace= True)
df_titanic_2

PassengerId  Survived  Pclass  \
755          756         1       2   
469          470         1       3   
644          645         1       3   
78            79         1       2   
831          832         1       2   
..           ...       ...     ...   
453          454         1       1   
52            53         1       1   
796          797         1       1   
698          699         0       1   
599          600         1       1   

                                             Name     Sex    Age  SibSp  \
755                     Hamalainen, Master. Viljo    male   0.67      1   
469                 Baclini, Miss. Helene Barbara  female   0.75      2   
644                        Baclini, Miss. Eugenie  female   0.75      2   
78                  Caldwell, Master. Alden Gates    male   0.83      0   
831               Richards, Master. George Sibley    male   0.83      1   
..                                            ...     ...    ...    ...   
453                      Goldenberg, Mr. Samuel L    male  49.00      1   
52       Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.00      1   
796                   Leader, Dr. Alice (Farnham)  female  49.00      0   
698                      Thayer, Mr. John Borland    male  49.00      1   
599  Duff Gordon, Sir. Cosmo Edmund ("Mr Morgan")    male  49.00      1   

     Parch    Ticket      Fare Embarked  
755      1    250649   14.5000        S  
469      1      2666   19.2583        C  
644      1      2666   19.2583        C  
78       2    248738   29.0000        S  
831      1     29106   18.7500        S  
..     ...       ...       ...      ...  
453      0     17453   89.1042        C  
52       0  PC 17572   76.7292        C  
796      0     17465   25.9292        S  
698      1     17421  110.8833        C  
599      0  PC 17485   56.9292        C  

[415 rows x 11 columns]

In [24]:
# Converting dataframe to csv file

df_titanic_2.to_csv('df_titanic_2.csv', header= True)

In [25]:
# Converting dataframe to json file

df_passenger_info.to_json('df_passenger_info.json', orient='columns')

# NOTEBOOK END